libs/auth/src/cast_vote_record_hashes.ts

```diff
@@ -0,0 +1,308 @@
+import { Buffer } from 'buffer';
+import fs from 'fs/promises';
+import { sha256 } from 'js-sha256';
+import path from 'path';
+import { groupBy } from '@votingworks/basics';
+import { Client } from '@votingworks/db';
+
+/**
+ * A representation of a file
+ */
+export interface File {
+  fileName: string;
+  fileContents: string | Buffer;
+}
+
+/**
+ * Computes a hash for a single cast vote record, incorporating the cast vote record's directory
+ * name, file names, and file contents. The directory summary that we hash specifically mirrors the
+ * output of:
+ *
+ * find <directoryName> -type f | sort | xargs sha256sum
+ */
+export function computeSingleCastVoteRecordHash({
+  directoryName,
+  files,
+}: {
+  directoryName: string;
+  files: File[];
+}): string {
+  const filesSorted = [...files].sort((file1, file2) =>
+    file1.fileName.localeCompare(file2.fileName)
+  );
+  const fileHashes: string[] = [];
+  for (const { fileName, fileContents } of filesSorted) {
+    fileHashes.push(
+      `${sha256(fileContents)}  ${path.join(directoryName, fileName)}\n`
+    );
+  }
+  const directorySummary = fileHashes.join('');
+  return sha256(directorySummary);
+}
+
+/**
+ * The input to {@link computeCombinedHash}
+ */
+export type HashesToCombine = Array<{ hash: string; sortKey: string }>;
+
+/**
+ * Sorts the provided hashes using the specified sort key, concatenates the hashes, and hashes the
+ * result, yielding a combined hash
+ */
+export function computeCombinedHash(hashesToCombine: HashesToCombine): string {
+  return sha256(
+    [...hashesToCombine]
+      .sort((entry1, entry2) => entry1.sortKey.localeCompare(entry2.sortKey))
+      .map((entry) => entry.hash)
+      .join('')
+  );
+}
+
+//
+// Functions for managing cast vote record hashes in a SQLite DB table. This code is relevant for
+// export-time hash computation.
+//
+
+/**
+ * A schema for a Merkle tree (https://en.wikipedia.org/wiki/Merkle_tree) that allows for efficient
+ * hashing of continuously exported cast vote records
+ */
+export const CAST_VOTE_RECORD_HASHES_TABLE_SCHEMA = `
+create table cvr_hashes (
+  cvr_id_level_1_prefix text not null,
+  cvr_id_level_2_prefix text not null,
+  cvr_id text not null,
+  cvr_hash text not null
+);
```

Comment:
Where's the level 1, level 2 requirement / strategy coming from? Why any levels? Why 2? I don't see anything clearly in the wikipedia article.

Reply:
nvm I see it in the tech spec now, forgot about that document

Reply:
but it doesn't explain _why_

Reply:
Yeah great q that nudged me to rethink things a bit!

So ultimately, I think that more important than the number of levels is the branching factor at each level. The latter informs the former. The number of levels just determines how many hashes you have to compute every time you add something to the tree. The difference between computing 2 hashes and 3 hashes should be negligible, assuming that each hash doesn't involve too much data. Speaking of which, the amount of data needed as input to each hash, on average, is determined by the branching factor.

With our current setup, we branch up to 256 times from the root to level 1 (because 16 hex chars * 16 hex chars = 256 combos of 2 hex chars) and up to 256 times from level 1 to level 2. Assuming an upper bound of, say, 10,000 CVRs on a machine, we'll almost certainly fill out level 1 of the tree, i.e. take up all 256 branches. On the other hand, we likely won't fill out level 2 at any point, because 10,000 / 256 = ~39, which is significantly less than 256. So level 2 isn't going to do much for us.

Once we hit 10,000 CVRs, we'll need to perform 3 hash computations (not including the hash of the individual CVR itself) - the level 2 hash, level 1 hash, and root hash. On average, the root hash will involve retrieving and hashing 256 level 1 hashes, the level 1 hash will involve retrieving and hashing 39 level 2 hashes, and the level 2 hash will involve retrieving and hashing 1 CVR hash. 256 + 39 + 1 = 296 hash retrievals.

Thinking this through again, I'm considering keeping 2 levels but decreasing the branching factor at each from 256 to 16, by having the level 1 prefix be just the first char of the CVR ID and the level 2 prefix be just the first two chars of the CVR ID. 10,000 / 16 = 625, and 625 / 16 = ~39. So once we hit 10,000 CVRs, on average, the root hash will involve retrieving and hashing 16 level 1 hashes, the level 1 hash will involve retrieving and hashing 16 level 2 hashes, and the level 2 hash will involve retrieving and hashing 39 CVR hashes. 16 + 16 + 39 = 71 hash retrievals.

Going to work on this change now! But also happy to discuss further in person since I know this response is dense haha

Reply:
https://github.com/votingworks/vxsuite/pull/3898/commits/6666b155158b098d329d2db647fde1b36a99ce33

Reply:
Reading this again this morning and want to change my wording slightly here:

> So ultimately, I think that more important than the number of levels is the branching factor at each level. The latter informs the former.

I actually think that they're equally important and inform each other. Their combination determines how much data you have to retrieve and hash when adding a new entry to the tree. This can almost certainly be expressed with a formula involving a logarithm haha.